src/train.py

```diff
@@ -97,36 +97,100 @@ def speech_loss(gold_mel, stop_label, pred_mel, stop_pred):
     stop_loss = F.binary_cross_entropy_with_logits(stop_pred.squeeze(), stop_label)
     return pred_loss + stop_loss
 
+
+def cross_entropy(input, target, size_average=True):
+    """ Cross entropy that accepts soft targets
+    Args:
+         pred: predictions for neural network
+         targets: targets, can be soft
+         size_average: if false, sum is returned instead of mean
+
+    Examples::
+
+        input = torch.FloatTensor([[1.1, 2.8, 1.3], [1.1, 2.1, 4.8]])
+        input = torch.autograd.Variable(out, requires_grad=True)
+
+        target = torch.FloatTensor([[0.05, 0.9, 0.05], [0.05, 0.05, 0.9]])
+        target = torch.autograd.Variable(y1)
+        loss = cross_entropy(input, target)
+        loss.backward()
+    """
+    logsoftmax = nn.LogSoftmax(1)
+    if size_average:
+        return torch.mean(torch.sum(-target * logsoftmax(input), dim=1))
+    else:
+        return torch.sum(torch.sum(-target * logsoftmax(input), dim=1))
+
 def discriminator_loss(output, target):
-    return F.cross_entropy(output, target)
+    return cross_entropy(output, target)
+
+def discriminator_target(output, target_type, smoothing=0.1):
+    """
+    Create the target labels with smoothing
+    Params
+        -output : [batch_size x 2]
+        -type : 'text' or 'speech'
+        -smoothing : label smoothing factor
+    Return
+        [batch_size x 2] tensor with target smoothed labels
+    """
+    target = torch.zeros_like(output)
+    if target_type == 'text':
+        smoothed_target_label = [1 - smoothing, smoothing]
+    else:
+        # target_type == 'speech' implicitly
+        smoothed_target_label = [smoothing, 1 - smoothing]
+    target[:,] = smoothed_target_label
+    return target
+
 
 
 #####----- Use these to run a task on a batch ----#####
-def autoencoder_step(model, batch):
+def autoencoder_step(model, batch, use_dis_loss=False):
     """
     Compute and return the loss for autoencoders
     """
     x, y = process_batch(batch)
     character, mel, mel_input, _  = x
     gold_char, gold_mel, gold_stop = y
 
-    text_pred = model.text_ae(character).permute(1, 2, 0)
-    pred, stop_pred = model.speech_ae(mel, mel_input)
+    if use_dis_loss:
+        text_pred, t_hid = model.text_ae(character, use_dis_loss)
+        text_pred = text_pred.permute(1, 2, 0)
+        t_d_loss = discriminator_hidden_to_loss(model, t_hid, 'speech')
```

Comment:
Wait, this is the text autoencoder... shouldn't it be 'text'?

Reply:
The encoders should try to fool the discriminator, thus target should be 'speech'.

Reply:
Ah, got it since this is adversarial right